ykrt/src/mt.rs

```diff
@@ -347,12 +355,76 @@ impl MTThread {
                     }
                 }
             }
-            PHASE_TRACING => {
+            PHASE_TRACING | PHASE_TRACING_LOCK => {
+                if lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                    // PHASE_TRACING_LOCK requires special handling, because another thread might
+                    // have grabbed it on this Location even though this thread is tracing this
+                    // location.
+
+                    if self.inner.tracer.is_none() {
+                        // If this thread isn't tracing anything, there's little point in spinning
+                        // as the thread tracing this location is still plausibly alive.
+                        return None;
+                    }
+
+                    // We wait to grab the lock.
+                    // FIXME: this is a terrible spinlock.
+                    while lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                        lp = loc.state.load(Ordering::Acquire);
+                        yield_now();
+                    }
+
+                    if lp & PHASE_TAG != PHASE_TRACING {
+                        // While waiting for the Location to move beyond PHASE_TRACING_LOCK, we
+                        // probably raced with the thread tracing this location.
+                        return None;
+                    }
+                }
+
+                // Before we can do anything with this phase's value, we need to transition to
+                // `PHASE_TRACING_LOCK`.
+                loop {
+                    let old_lp = lp;
+                    lp = loc
+                        .state
+                        .compare_and_swap(lp, PHASE_TRACING_LOCK, Ordering::Acquire);
+                    if lp == old_lp {
+                        break;
+                    }
+                    if self.inner.tracer.is_none() {
+                        // This thread isn't tracing anything, so it's not worth us spinning and
+                        // trying to lock this location.
+                        return None;
+                    }
+                    // We are tracing something, and since it might be the current Location, we
+                    // need to transition to TRACING_LOCK to avoid tracing more than one iteration
```

Comment:
I thought we'd already transitioned to `PHASE_TRACING_LOCK` above?

Reply:
At this point we failed to transition to `PHASE_TRACING_LOCK` so we need to try again (or give up).